### PR TITLE
[sql lab] a better approach at limiting queries

### PR DESCRIPTION
**Pull request by mistercrunch**
Currently there are two mechanisms that we use to enforce the row
limiting constraints, depending on the database engine:
1. use dbapi's `cursor.fetchmany()`
2. wrap the SQL into a limiting subquery

Method 1 isn't great as it can result in the database server storing
larger than required result sets in memory expecting another fetch
command while we know we don't need that.

Method 2 has a positive side of working with all database engines,
whether they use LIMIT, ROWNUM, TOP or whatever else since sqlalchemy
does the work as specified for the dialect. On the downside though
the query optimizer might not be able to optimize this as much as an
approach that doesn't use a subquery.

Since most modern DBs use the LIMIT syntax, this adds a regex approach
to modify the query and force a LIMIT clause without using a subquery
for the database that support this syntax and uses method 2 for all
others.

### CODE REVIEW

File: superset/db_engine_specs.py [hunk 1]
```diff
@@ -88,6 +90,23 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+            qry = (
+                select('*')
+                .select_from(
+                    TextAsFrom(text(sql), ['*']).alias('inner_qry'),
+                )
+                .limit(limit)
+            )
+            return database.compile_sqla_query(qry)
+        elif LimitMethod.FORCE_LIMIT:
+            no_limit = re.sub(r'(?i)\s+LIMIT\s+\d+;?(\s|;)*$', '', sql)
```
Comment (betodealmeida) [inline superset/db_engine_specs.py]: Scary! :-O

BTW, you can use backreferences here to replace the LIMIT number in one pass (untested):

```python
sql = re.sub(r'(?i)(.*\s+LIMIT\s+)\d+\s*;?\s*$', r'\1 {0}'.format(limit), sql)
```

File: superset/models/core.py [hunk 2]
```diff
@@ -722,14 +720,7 @@ def select_star(
...
     def wrap_sql_limit(self, sql, limit=1000):
-        qry = (
-            select('*')
-            .select_from(
-                TextAsFrom(text(sql), ['*'])
-                .alias('inner_qry'),
-            ).limit(limit)
-        )
-        return self.compile_sqla_query(qry)
+        return self.db_engine_spec.apply_limit_to_sql(sql, limit, self)
```
Comment (betodealmeida) [inline superset/models/core.py]: Better update the method name here.

File: tests/db_engine_specs_test.py [hunk 3]
```diff
@@ -80,3 +81,45 @@ def test_job_2_launched_stage_2_stages_progress(self):
...
+        limited = MySQLEngineSpec.apply_limit_to_sql(sql, 1000, db)
+        expected = 'SELECT * FROM a LIMIT 1000'
+        self.assertEquals(expected, limited)
+
+    def test_limit_query_with_limit_subquery(self):
+        sql = 'SELECT * FROM (SELECT * FROM a LIMIT 10) LIMIT 9999'
+        db = Database(sqlalchemy_uri='mysql://localhost')
+        limited = MySQLEngineSpec.apply_limit_to_sql(sql, 1000, db)
+        expected = 'SELECT * FROM (SELECT * FROM a LIMIT 10) LIMIT 1000'
+        self.assertEquals(expected, limited)
```
Comment (betodealmeida) [inline tests/db_engine_specs_test.py]: I'm nervous about the regular expression, I'd add a few more unit tests covering some weird cases, eg:

```sql
SELECT
    'LIMIT 1000' AS a
  , b
FROM
    table
LIMIT
    1000
<EOF>
```

And maybe some stuff with extra space at the end:

```sql
...
LIMIT      1000  ;    
  <EOF>
```

**Review by betodealmeida — COMMENTED** [review]
LGTM, but I have a few comments.

File: superset/models/core.py (hunk 2 above)
Comment (betodealmeida) [reply superset/models/core.py]: I mean, rename `wrap_sql_limit` to something else.

File: superset/db_engine_specs.py (hunk 1 above)
Comment (mistercrunch) [reply superset/db_engine_specs.py]: I barely understand what I wrote here :/

Comment (betodealmeida) [reply superset/db_engine_specs.py]: > Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems.

Yeah, I just realized my example does not work when there's no LIMIT clause. :-/

You know about verbose regexes? Eg:

```python
no_limit = re.sub(r"""
    (?ix)        # case insensitive, verbose
    \s+          # whitespace
    LIMIT\s+\d+  # LIMIT $ROWS
    ;?           # optional semi-colon
    (\s|;)*$     # any number of trailing whitespace or semi-colons, til the end
""", '', sql)
```

Though in this case I think it's pretty straightforward. You might wanna `rstrip` whitespace and semicolons from the end of `sql`, that would simplify the regex, no?

File: superset/models/core.py (hunk 2 above)
Comment (mistercrunch) [reply superset/models/core.py]: went with `apply_limit_to_sql`

**Comment by mistercrunch** [comment]
Addressed comments, merging

**Comment by mistercrunch** [comment]
I think we agree that we need a way to prevent users from running infinite billion rows query that will hurt the db server, network, web server and ultimately crash their browser. 

Now we've been doing this two ways depending on the DB engine, `fetchmany`, which I think puts too much strain on the database server. It's rough on Presto/Hive for instance and in many cases is a similar internal cost as running the whole thing from the db's perspective. Then there's the wrap-in-a-subquery approach, that's better and prefer that to `fetchmany`, though I think many db optimizer won't do as good with that as with the `LIMIT` in the main core query. It requires more complex query editing but is preferable to the subquery approach as it may result in a simpler query plan.

I also looked at how other tools seem to be doing this, and they edit the user query much like we do here, likely for the reasons listed above.

**Pull request merged**